src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs

```diff
@@ -55,36 +55,39 @@ private IDictionary CreateDataContainer()
         [MethodImpl(MethodImplOptions.InternalCall)]
         private static extern bool IsImmutableAgileException(Exception e);
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern IRuntimeMethodInfo GetMethodFromStackTrace(object stackTrace);
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "ExceptionNative_GetMethodFromStackTrace")]
+        private static partial void GetMethodFromStackTrace(ObjectHandleOnStack stackTrace, ObjectHandleOnStack method);
 
         private MethodBase? GetExceptionMethodFromStackTrace()
         {
-            Debug.Assert(_stackTrace != null, "_stackTrace shouldn't be null when this method is called");
-            IRuntimeMethodInfo method = GetMethodFromStackTrace(_stackTrace!);
+            if (_stackTrace == null)
+            {
+                return null;
+            }
+
+            object stackTraceLocal = _stackTrace;
+            IRuntimeMethodInfo? methodInfo = null;
+            GetMethodFromStackTrace(ObjectHandleOnStack.Create(ref stackTraceLocal), ObjectHandleOnStack.Create(ref methodInfo));
+
+            // Ensure the stack trace lives beyond the current above call.
+            // See implementation for details on stacktrace data.
+            GC.KeepAlive(stackTraceLocal);
 
             // Under certain race conditions when exceptions are re-used, this can be null
-            if (method == null)
+            if (methodInfo == null)
+            {
                 return null;
+            }
 
-            return RuntimeType.GetMethodBase(method);
+            return RuntimeType.GetMethodBase(methodInfo);
         }
 
         public MethodBase? TargetSite
         {
             [RequiresUnreferencedCode("Metadata for the method might be incomplete or removed")]
             get
             {
-                if (_exceptionMethod != null)
-                {
-                    return _exceptionMethod;
-                }
-                if (_stackTrace == null)
-                {
-                    return null;
-                }
-
-                _exceptionMethod = GetExceptionMethodFromStackTrace();
+                _exceptionMethod ??= GetExceptionMethodFromStackTrace();
                 return _exceptionMethod;
             }
```

Comment:
```suggestion
            get => _exceptionMethod ??= GetExceptionMethodFromStackTrace();
```
Nit: This can be a one-liner.